[jit] Add `clone_instance` for Module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30168 [jit] Add `clone_instance` for Module**

Summary:
Previous implementation of `clone` in `script::Module` copies both the module instance and the
class type, after we enabled type sharing https://github.com/pytorch/pytorch/pull/26666 we also
need to have a function to clone instance only and share the underlying class type.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18631324](https://our.internmc.facebook.com/intern/diff/D18631324)

jerryzh168 committed:
[jit] Add `clone_instance` for Module

Summary:
Previous implementation of `clone` in `script::Module` copies both the module instance and the
class type, after we enabled type sharing https://github.com/pytorch/pytorch/pull/26666 we also
need to have a function to clone instance only and share the underlying class type.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Add `clone_instance` for Module"


Summary:
Previous implementation of `clone` in `script::Module` copies both the module instance and the
class type, after we enabled type sharing https://github.com/pytorch/pytorch/pull/26666 we also
need to have a function to clone instance only and share the underlying class type.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/jit/test_module_api.cpp:
```diff
@@ -0,0 +1,45 @@
+  Module m3 = m.clone_instance();
+  // Make sure copy works
+  ASSERT_TRUE(m2.attr(attr_name).toInt(), 2);
+  ASSERT_TRUE(m3.attr(attr_name).toInt(), 2);
```
Should it be `ASSERT_EQ`?

ZolotukhinM commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -237,6 +237,29 @@ Module Module::clone_impl(
+    if (type()->getAttribute(i)->is_module()) {
+      const Module& orig = Module(s.toObject());
+      Module cloned = orig.clone_instance();
+      r.register_module(type()->getAttributeName(i), cloned);
```
It would be better to not use `register_*` API here and explicitly use `setattr` on the object itself. The `clone_instance` function should not be able to mutate the type, but `register_`* is a mutating API. I realize that it doesn't change it in this particular case, but it still would be cleaner not to use it at all. Think about the type as having `const` modifier in this function and `register_` methods are not marked `const` so we're getting a compilation error here :)

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -224,6 +224,7 @@ struct TORCH_API Module : public Object {
 
   // Create a deep copy of this module.
   Module clone() const;
+  Module clone_instance() const;
```
Please add comments for both methods. The existing comment is not very helpful.

ZolotukhinM reviewed (approved):
Overall looks good. Do you want to also add python bindings for it?

jerryzh168 commented:
> Overall looks good. Do you want to also add python bindings for it?

will do

jerryzh168 committed:
Update on "[jit] Add `clone_instance` for Module"


Summary:
Previous implementation of `clone` in `script::Module` copies both the module instance and the
class type, after we enabled type sharing https://github.com/pytorch/pytorch/pull/26666 we also
need to have a function to clone instance only and share the underlying class type.

Test Plan:
tbd

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
